### PR TITLE
[4.0] Translate stages via language files

### PR DESCRIPTION
**Pull request by bembelimen**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/21476#issuecomment-411694584.

### Summary of Changes
Now one can use languages constants to have multilingual states

### Testing Instructions
Create a state called "JPUBLISHED" and check it everywhere in the system (state list, article edit form, article list, transition edit form, ...)

### Expected result
The language constant will be translated

### Actual result
No translation at all

### CODE REVIEW

File: administrator/components/com_workflow/tmpl/states/default.php [hunk 1]
```diff
@@ -11,8 +11,9 @@
  */
 defined('_JEXEC') or die;
 
-use Joomla\CMS\Language\Text;
+use Joomla\CMS\Factory;
```
Comment (brianteeman) [inline administrator/components/com_workflow/tmpl/states/default.php]: Factory already exists a few lines down. Adding it twice will fail.

Comment (bembelimen) [reply administrator/components/com_workflow/tmpl/states/default.php]: Strange, I thought I solved it with the conflict, thanks for the pointer!

File: administrator/components/com_workflow/tmpl/states/default.php [hunk 2]
```diff
@@ -120,10 +122,10 @@
 										<?php if ($canEdit) : ?>
 											<?php $editIcon = '<span class="fa fa-pencil-square mr-2" aria-hidden="true"></span>'; ?>
 											<a href="<?php echo $edit; ?>" title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape(addslashes($item->title)); ?>">
```
Comment (infograf768) [inline administrator/components/com_workflow/tmpl/states/default.php]: need also a Text:: for the title
<?php echo $this->escape(addslashes($item->title))
to
<?php echo $this->escape(addslashes(Text::_($item->title)))

**Comment by infograf768** [comment]
Also:
same corrections should be done for Transitions
Using maybe the existing strings
JTOOLBAR_UNPUBLISH, etc. or new specific COM_WORKFLOW strings

To obtain (here for French. just modified for unpublished)

screen shot 2018-08-10 at 09 02 43

In both cases `Modifier Non Publié` for the title is not very nice phrasing. Maybe add a : between.

**Comment by bembelimen** [comment]
@infograf768 I'll create a new PR for this, but thanks for the suggestion!

**Comment by infograf768** [comment]
Can conflicts be solved?
And obviously adapted to the new stage name instead of state

**Comment by bembelimen** [comment]
Sure, I waited for #21566

**Comment by infograf768** [comment]
I think that also `Joomla! Default` should be a string btw.

File: libraries/src/Form/Field/WorkflowStageField.php [hunk 3]
```diff
@@ -124,7 +126,7 @@ protected function getGroups()
 				$workflowStages[$workflowStageKey] = array();
 			}
 
-			$workflowStages[$workflowStageKey][] = HTMLHelper::_('select.option', $stage->workflow_stage_id, $stage->workflow_stage_title);
+			$workflowStates[$workflowStateKey][] = HTMLHelper::_('select.option', $state->workflow_stage_id, Text::_($state->workflow_stage_title));
```
Comment (infograf768) [inline libraries/src/Form/Field/WorkflowStageField.php]: $state or $stage?

**Comment by infograf768** [comment]
Also conflict with new file name.

**Comment by infograf768** [comment]
And I foresee issues with https://github.com/joomla/joomla-cms/pull/21520 as the sql does not use string constants.

File: libraries/src/Form/Field/TransitionField.php [hunk 4]
```diff
@@ -13,6 +13,9 @@
 use Joomla\CMS\Factory;
 use Joomla\CMS\Form\FormHelper;
 use Joomla\CMS\Workflow\Workflow;
+use Joomla\CMS\Form\Field\ListField;
```
Comment (Joomv) [inline libraries/src/Form/Field/TransitionField.php]: You don't need to import this as it's in the same namespace

File: libraries/src/Form/Field/WorkflowStageField.php (hunk 3 above)
Comment (Joomv) [reply libraries/src/Form/Field/WorkflowStageField.php]: Should be $stage

File: administrator/components/com_workflow/tmpl/stages/default.php [hunk 5]
```diff
@@ -23,6 +23,8 @@
 $user      = Factory::getUser();
 $userId    = $user->id;
 
+$lang		= Factory::getLanguage();
```
Comment (Joomv) [inline administrator/components/com_workflow/tmpl/stages/default.php]: You aren't using this property

**Review by Joomv — CHANGES_REQUESTED** [review]
Also requires conflicts fixing as well as comments

Deleted useless hashtag here... (JM)

File: libraries/src/Form/Field/WorkflowStageField.php (hunk 3 above)
Comment (bembelimen) [reply libraries/src/Form/Field/WorkflowStageField.php]: Fixed

**Comment by infograf768** [comment]
Missing Text:: line 118 of tmpl/transitions/default.php to get
`											<a href="<?php echo $edit; ?>" title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo Text::_($this->escape(addslashes($item->title))); ?>">
`

Otherwise good first step for me

File: administrator/components/com_content/tmpl/articles/default.php [hunk 6]
```diff
@@ -222,7 +222,7 @@
 											<span class="icon-<?php echo $icon; ?>"></span>
 										<?php endif; ?>
 										</div>
-										<div class="mr-auto"><?php echo $this->escape($item->stage_title); ?></div>
+										<div class="mr-auto"><?php echo Text::_($this->escape($item->stage_title)); ?></div>
```
Comment (Fedik) [inline administrator/components/com_content/tmpl/articles/default.php]: I think correct would be do `escape()` after translation, not before :wink:

Comment (bembelimen) [reply administrator/components/com_content/tmpl/articles/default.php]: You're right, thx

**Comment by bembelimen** [comment]
Thanks @infograf768

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 9d0ba736eb75e0142a719a7b5605fd7cbe153507<br /><br />Let's advance with this.

One more tester, if possible.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21506">issues.joomla.org/tracker/joomla-cms/21506</a>.</sub>

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 9d0ba736eb75e0142a719a7b5605fd7cbe153507<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21506">issues.joomla.org/tracker/joomla-cms/21506</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by alikon** [comment]
https://github.com/joomla/joomla-cms/pull/21506/files#diff-135c6f58583408312a709459b19594c7
it lacks the upgrade path and obviously postgresql

done in https://github.com/bembelimen/joomla-cms/pull/22

**Comment by alikon** [comment]
question:
should we do the same on the `#__workflow_transitions` ?

**Comment by ghost** [comment]
Status back on "Pending".

**Comment by infograf768** [comment]
@alikon 
I thought we could do that after the follow-up on this.
I explain:

We also need to translate the transitions. I.e. enter in db some string for the actions
Unpublish, Publish, etc.
Then translate the titles in transitions Manager with something like
```
										<?php echo $editIcon; ?><?php echo Text::_($item->title); ?>
											</a>
										<?php else: ?>
											<?php echo Text::_($item->title); ?>
										<?php endif; ?>
```
so basically we would have something like
```
INSERT INTO `#__workflow_transitions` (`id`, `asset_id`, `published`, `ordering`, `workflow_id`, `title`, `description`, `from_stage_id`, `to_stage_id`) VALUES
(1, 0, 1, 1, 1, 'JUNPUBLISH', '', -1, 1),
(2, 0, 1, 2, 1, 'JPUBLISH', '', -1, 2),
(3, 0, 1, 3, 1, 'JTRASH', '', -1, 3),
(4, 0, 1, 4, 1, 'JARCHIVE', '', -1, 4);
```
And add the new lang strings.
Is it worth changing now db or do it in the next PR?

**Comment by infograf768** [comment]
Or do it in this PR?

**Comment by alikon** [comment]
Yes we can do in a next PR,

Il ven 24 ago 2018, 07:28 infograf768 <notifications@github.com> ha scritto:

**Comment by infograf768** [comment]
@bembelimen 
Your decision. If you do transitions in next PR, let's not forget @alikon necessary updated patch.

**Comment by bembelimen** [comment]
Thanks @alikon for the PR

I would do the translation in a new PR, so we have small code changes.

**Comment by infograf768** [comment]
Drone error is unrelated to this PR.

@wilsonge Please merge.

**Comment by infograf768** [comment]
rtc
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/21506">issues.joomla.org/tracker/joomla-cms/21506</a>.</sub>

**Comment by brianteeman** [comment]
Whoever merges can you directly fix the drone error

**Comment by wilsonge** [comment]
Thanks guys!

**Pull request merged**